---
title: 'Lin et al.: AAV Microglia'
subtitle: 'Preprocessing of 10X scRNA-seq Reference Dataset (Supplementary Figure 2)'
author: "Ruiyu Ray Wang"
date: "`r Sys.Date()`"
output: html_document
knit: (function(input_file, encoding) {
    out_dir <- '../docs';
    rmarkdown::render(input_file,
      encoding=encoding,
      output_file=file.path(dirname(input_file), out_dir, 'FigS2_scRNAseq_reference_QC.html'))})
---

```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## TL;DR

You may click on the following links to fast forward to the codes directly producing the Figures:

  - [Supplementary Figure 2a](#FigS2a)
  - [Supplementary Figure 2b](#FigS2b)
  - [Supplementary Figure 2c](#FigS2c)
  - [Supplementary Figure 2d](#FigS2d)

Back to [index page](https://ruiyuraywang.github.io/Lin_AAV_Microglia_public/index.html).

## About this Vignette

This vignette illustrates the steps by which the reference data (10X) was processed. It also reproduces some of the Supplementary Figures in the original publication involved with the pre-processing of reference data (Lin et al., 2022).  
Here we only show how plain plots were generated directly from data. To reach the final illustrations in the publication, these plots may require further editing in graphic design softwares (i.e. Adobe Illustrator). Those beautification steps are not central to our analysis and therefore will be omitted.

The input file of this vignette is a filtered `h5` file generated by `CellRanger` (`GSE197743_AAV_MG_filtered_feature_bc_matrix.h5`). The output is a processed `.h5Seurat` with metadata slot filled.

### Overview of the experiments

Microglia were isolated from adult mice (~8w) using a method based on the "cold-mechanical dissociation" protocol (Bennett et al., 2016)[^1]. The steps involving FACS sorting from the original protocol was omitted to minimize *ex vivo* activation of microglia.

We found that the single cell transcriptome acquired by 10X Chromium (3' Chemistry v3) captured transcripts of the AAV-packaged transgene (mScarlet) at a very low rate (data not published). Therefore we carried out deep single cell sequencing using a modified protocol based on a customized Smart-seq2 protocol (Zhang et al., 2020)[^2] on microglia isolated from mice after stereotaxic AAV delivery. The Smart-seq2 based method captured mScarlet transcripts with substantially higher efficiently.

To characterize the microglia transcriptomes under quiescent and immune activation states, we isolated microglia from mice subjected to Saline, Poly(i:c) and LPS treatment (i.p.), and performed scRNA-seq using the 10X Chromium platform.

### Analysis Outline

Upstream analysis pipelines (i.e. sequence alignment, generation of digital expression matrix, etc.):

  * For Smart-seq2, fastq files were processed by a [custom Snakemake workflow](https://github.com/RuiyuRayWang/ScRNAseq_smkpipe_at_Luolab) to generate the expression table (feature-barcode-count matrix).
  * For 10X, fastq files from 3 mice were quantified by CellRanger (v6.0.1) using the `cellranger count` command and then aggregated using the `cellranger aggr` command.

Downstream analysis:

  * Preprocessing and QC (`scater`)
  * Seurat Standard Workflow (`Seurat`)
  * Label Transfer (`Seurat`)

## Pre-processing

### Load libraries and data

Load reuiqred libraries.
```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(SeuratDisk)
  library(scater)
  library(tidyverse)
  library(RColorBrewer)
  library(scales)
})
```

If you haven't done so, download data from GEO.
```{r}
if(!file.exists('../data/GSE197743/GSE197743_AAV_MG_filtered_feature_bc_matrix.h5')){
  library(GEOquery)
  options(timeout = 6000)  ## In case of slow Internet connections
  getGEOSuppFiles(GEO = "GSE197743", makeDirectory = T, baseDir = '../data/')
}
```

Load the data matrix to workspace.
```{r}
data <- Read10X_h5(filename = '../data/GSE197743/GSE197743_AAV_MG_filtered_feature_bc_matrix.h5')
```
Create Seurat object.
```{r}
cells <- CreateSeuratObject(data, project = "AAV_Microglia", names.field = 2, names.delim = "-")
```

Update metadata.
```{r}
Idents(cells) <- "orig.ident"
cells <- RenameIdents(cells, `1` = "MG_S", `2` = "MG_P", `3` = "MG_L")
cells[["group"]] <- Idents(cells)

Idents(cells) <- "orig.ident"
cells <- RenameIdents(cells, `1` = "Saline", `2` = "Poly(i:c)", `3` = "LPS")
cells[["treat"]] <- Idents(cells)

cells[["id"]] <- "Reference"
cells[["tech"]] <- "10x"
```

### Quality Control (QC)

We use the `scater`[^3] package for quality control (QC) of the single cell datasets
```{r}
cells <- DietSeurat(cells)
cells.sce <- as.SingleCellExperiment(cells)
cells.sce <- logNormCounts(cells.sce)
dim(cells.sce)
```

#### Cell level QC

```{r}
per.cell <- perCellQCMetrics(cells.sce,
                             subsets = list(Mito = grep("^mt-", rownames(cells.sce)),  # Mitochondrial genes
                                            Ribo = grep("\\bRp[sl]\\d+[^k]*\\b", rownames(cells.sce))))  # Ribosomal genes
qc.stats <- quickPerCellQC(per.cell, percent_subsets = c("subsets_Mito_percent","subsets_Ribo_percent"))

qc.stats$high_subsets_Mito_percent <- per.cell$subsets_Mito_percent > 3
qc.stats$discard <- qc.stats$low_lib_size | qc.stats$low_n_features | qc.stats$high_subsets_Mito_percent | qc.stats$high_subsets_Ribo_percent
qc.stats$discard_size_complexity <- qc.stats$low_lib_size | qc.stats$low_n_features
colData(cells.sce) <- cbind(colData(cells.sce), cbind(per.cell, qc.stats))
colSums(as.matrix(qc.stats))
```
```{r}
cells.filtered <- cells.sce[,!qc.stats$discard]
```

#### Feature level QC

```{r}
keep_feature <- nexprs(cells.filtered, byrow=TRUE) > 0

## Remove mitochondrial and ribosomal genes which does not contribute to dimensionality reduction and clustering
keep_feature[grep("\\bRp[sl]\\d+[^k]*\\b", rownames(cells.sce))] <- FALSE
keep_feature[grep("mt-", rownames(cells.sce))] <- FALSE
cells.filtered <- cells.filtered[keep_feature,]
```

Back to Seurat object.
```{r}
refdata <- as.Seurat(cells.filtered)
refdata[["nCount_originalexp"]] <- NULL; refdata[["nFeature_originalexp"]] <- NULL
refdata <- RenameAssays(refdata, originalexp = "RNA")
refdata$ident <- NULL
```

### Standard Seurat Workflow

```{r}
refdata <- refdata %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA(npcs = 50)

## USE DIM=1:50 FOR UNFILTERED DATA, SKIP JACKSTRAW TO SAVE TIME
# refdata <- JackStraw(refdata, dims = 50)
# refdata <- ScoreJackStraw(refdata, dims = 1:50)
# JackStrawPlot(refdata, dims = 1:50)
# dims_use <- JS(refdata[["pca"]], slot = "overall") %>% as.data.frame %>% filter(Score < 0.05) %>% pull(PC)

dims_use <- 1:50

refdata <- refdata %>%
  FindNeighbors(dims = dims_use) %>%
  FindClusters(resolution = 0.8) %>%
  RunUMAP(dims = dims_use)
```

Take a closer look at our pre-processed data:  
We see that there are some major clusters in the middle and bottom-right of the umap plot. These are microglia single cell populations. The rest of the clusters represent non-microglia populations as well as droplets contaminated by cell debris. These were captured in our data because we applied Dounce homogenization and Percoll density gradient centrifugation to isolate microglia from mouse brain, which can never reach perfect purity.  
Nevertheless, the microglia clusters look good, as they exhibit robust `Hexb` expression, a well validate microglia marker[^4]. We can easily isolate them via *in silico* procedures, as is shown in subsequent code chunks.
```{r fig.width=10, fig.height=4}
FeaturePlot(refdata, features = "Hexb") | FeaturePlot(refdata, features = "subsets_Mito_percent")
```

Some microglia single cells exhibit high mitochondrial RNA levels, purge them.
```{r}
refdata_clean <- subset(refdata, subset = subsets_Mito_percent < 3)
```

To achieve *in silico* purification of microglia populations, we use the `CellSelector()` utility provided by `Seurat` package. This procedure requires manual selection of cells and is difficult to replicate programmatically.  
To help you reproduce the result, we have saved the identity of the selected cells in `../miscs/mg_manual.rda` in this repo. You can load it directly into workspace to subset out the micorglia population and proceed to downstream workflows.
```{r}
## NOT RUN
# Idents(refdata_clean) <- "contaminants"
# refdata_clean <- CellSelector(FeaturePlot(refdata_clean, features = "Hexb"), object = refdata_clean, ident = "microglia")  ## Run repeatedly, select desired clusters
# mg_manual <- WhichCells(refdata_clean, idents = "microglia")
# save(mg_manual, file = "../miscs/mg_manual.rda")
```
```{r}
load(file = '../miscs/mg_manual.rda')
refdata_clean <- subset(refdata_clean, cells = mg_manual)
```

Re-run Seurat Standard workflow on subsetted data.
```{r}
refdata_clean <- refdata_clean %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA(npcs = 45) %>%
  FindNeighbors(dims = 1:45) %>%
  FindClusters() %>%
  RunUMAP(dims = 1:45)
```

### Update Metadata

The metadata slot is updated based on the experiment design.  
But note that in recent years the microglia community is gradually quitting the use of `Reactive` and `Homeostatic` nomenclature.
```{r}
Idents(refdata_clean) <- "orig.ident"
refdata_clean <- RenameIdents(refdata_clean, `1` = "Homeostatic", `2` = "Reactive", `3` = "Reactive")
refdata_clean[["state"]] <- Idents(refdata_clean)
```

## Reproducing Plots

A general note on the UMAP visualizations:  
Depending on the use of CPUs and versions of the `umap` algorithm, the UMAP coordinates computed on different machines may not entirely match what was shown in the original publication. This is a known [issue](https://github.com/lmcinnes/umap/issues/525) of the `umap` algorithm. Nevertheless, the overall structure of the UMAP embedding should be consistent, i.e. microglia from mouse subjected to Saline, Poly(i:c) and LPS groups reside in different clusters. We've tested this on different computing platforms and reached the conclution that the shifts in UMAP coordinates does not affect our biological interpretations.

### Supplementary Figure 2a {#FigS2a}

```{r fig.width=5, fig.height=3.6}
as.data.frame(colData(cells.sce)[,c("nCount_RNA","nFeature_RNA","discard_size_complexity")]) %>%
  ggplot(mapping = aes(x=nCount_RNA, y=nFeature_RNA)) +
  geom_point(aes(color = discard_size_complexity), size = 1.2) +
  # ggtitle("Reference Data (10x)") +
  scale_color_manual(values = brewer.pal(n = 6,name = "Paired")[c(2,6)], labels = c("Kept","Removed"), name = "") +
  scale_x_continuous(expand = c(0.15,0.1)) +
  xlab("UMI counts") +
  ylab("Gene counts") +
  theme_classic() +
  theme(
    # plot.title = element_text(size = 12, face = "bold"),
    axis.text = element_text(size = 16),
    axis.title = element_text(size = 21),
    legend.title = element_blank(),
    legend.box.background = element_blank(),
    legend.position = c(.95, .05),
    legend.justification = c("right", "bottom"),
    legend.text = element_text(size = 16),
    legend.margin = margin(1, 5, 5, 5)
    )
```

### Supplementary Figure 2b {#FigS2b}

```{r fig.width=3, fig.height=3.6}
as.data.frame(colData(cells.sce)[,c("treat","subsets_Mito_percent","high_subsets_Mito_percent")]) %>%
  ggplot(mapping = aes(x=treat, y=subsets_Mito_percent)) +
  geom_violin() +
  geom_jitter(mapping = aes(color = high_subsets_Mito_percent), size = 0.3, height = 0, width = 0.1) +
  scale_color_manual(values = brewer.pal(n = 6,name = "Paired")[c(2,6)]) +
  xlab(NULL) +
  ylab("Mitochondrial RNA %") +
  theme_linedraw() +
  theme(
    legend.position = "none",
    axis.text.y = element_text(size = 14),
    axis.text.x = element_text(size = 16),
    axis.title = element_text(size = 20),
    )
```

### Supplementary Figure 2c {#FigS2c}

```{r fig.width=3, fig.height=3.6}
as.data.frame(colData(cells.sce)[,c("treat","subsets_Ribo_percent","high_subsets_Ribo_percent")]) %>%
  ggplot(mapping = aes(x=treat, y=subsets_Ribo_percent)) +
  geom_violin() +
  geom_jitter(mapping = aes(color = high_subsets_Ribo_percent), size = 0.3, height = 0, width = 0.1) +
  scale_color_manual(values = brewer.pal(n = 6,name = "Paired")[c(2,6)]) +
  xlab(NULL) +
  ylab("Ribosomal RNA %") +
  theme_linedraw() +
  theme(
    legend.position = "none",
    axis.text.y = element_text(size = 14),
    axis.text.x = element_text(size = 16),
    axis.title = element_text(size = 20),
    )
```

### Supplementary Figure 2d {#FigS2d}

```{r}
refdata_clean$treat <- factor(refdata_clean$treat, levels = c("Saline","Poly(i:c)","LPS"))
umap_reff1 <- FeaturePlot(refdata, features = "Hexb") +
  ggtitle("Hexb") +
  theme(
    plot.title = element_text(hjust = 0, size = 24),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    axis.title = element_text(size = 18),
    axis.line = element_line(arrow = grid::arrow(length = unit(0.3, "cm"), 
                                                       ends = "last", type = "closed")),
    legend.title = element_blank(),
    legend.text = element_text(size = 14),
    legend.position = c(.05, .25)
  )
umap_reff2 <- DimPlot(refdata, cells.highlight = mg_manual) + 
  NoLegend() +
  ggtitle("Retained Cells") +
  theme(
    plot.title = element_text(hjust = 0, size = 24),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    axis.title = element_text(size = 18),
    axis.line = element_line(arrow = grid::arrow(length = unit(0.3, "cm"), 
                                                 ends = "last", type = "closed"))
  )
umap_reff3 <- DimPlot(refdata_clean, group.by = "treat") +
  scale_color_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,3,5)], 
                     labels = c("Saline", "Poly(i:c)", "LPS")) +
  ggtitle("Final Reference") +
  guides(color = guide_legend(override.aes = list(size=3), byrow = TRUE)) +
  theme(
    plot.title = element_text(hjust = 0, size = 24),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    axis.title = element_text(size = 18),
    axis.line = element_line(arrow = grid::arrow(length = unit(0.3, "cm"), 
                                                 ends = "last", type = "closed")),
    legend.title = element_blank(),
    legend.text = element_text(size = 12),
    legend.position = c(.08, .22),
    legend.spacing.y = unit(0.25, 'cm')
  )
```
```{r fig.width=12, fig.height=4}
umap_reff1 | umap_reff2 | umap_reff3
```

## Save Results

Backup the processed Seurat object using `h5` format.
```{r}
## Avoid using factor while saving to other formats
refdata_clean$group <- as.character(refdata_clean$group)
refdata_clean$treat <- as.character(refdata_clean$treat)
refdata_clean$state <- as.character(refdata_clean$state)
refdata_clean$orig.ident <- as.character(refdata_clean$orig.ident)
refdata_clean$RNA_snn_res.0.8 <- as.integer(as.character(refdata_clean$RNA_snn_res.0.8))
refdata_clean$seurat_clusters <- as.integer(as.character(refdata_clean$seurat_clusters))

if(!dir.exists('../data/AAV_MG_scRNAseq')){dir.create('../data/AAV_MG_scRNAseq')}
SaveH5Seurat(refdata_clean, filename = '../data/AAV_MG_scRNAseq/refdata_clean.h5Seurat', overwrite = T, verbose = F)
Convert('../data/AAV_MG_scRNAseq/refdata_clean.h5Seurat', dest = "h5ad", overwrite = T, verbose = F)
```

### Session info

```{r}
sessionInfo()
```

[^1]: Bennett, M. L. et al. New tools for studying microglia in the mouse and human CNS. Proc Natl Acad Sci USA 113, E1738 (2016).
[^2]: Zhang, S. et al. Single-cell transcriptomics identifies divergent developmental lineage trajectories during human pituitary development. Nat Commun 11, 5275 (2020).
[^3]: McCarthy, D. J., Campbell, K. R., Lun, A. T. L. & Wills, Q. F. Scater: pre-processing, quality control, normalization and visualization of single-cell RNA-seq data in R. Bioinformatics btw777 (2017) doi:10.1093/bioinformatics/btw777.
[^4]: Masuda, T. et al. Novel Hexb-based tools for studying microglia in the CNS. Nat Immunol (2020) doi:10.1038/s41590-020-0707-4.

