---
title: 'Lin et al.: AAV Microglia'
author: "Ruiyu Ray Wang"
output: html_document
knit: (function(input_file, encoding) {
    out_dir <- '../docs';
    rmarkdown::render(input_file,
      encoding=encoding,
      output_file=file.path(dirname(input_file), out_dir, 'index.html'))})
---

```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## Introduction

Characterization of microglia polarization states under targeted AAV transduction using bulk and single cell RNA sequencing.

The study of microglia biology and the development of microglia-based gene therapies are in urgent need of efficient and safe vehicles for microglia transgene delivery. To address this, we developed adeno-associated virus (AAV) variants that mediate efficient in vitro and in vivo microglia transduction via directed evolution of the AAV capsid protein. To assess the effect of AAV transduction on microglia, we carried out bulk RNAseq in primary microglia and found that microglia transduced by AAV remain close to homeostatic state. Furthermore, single-cell RNA sequencing showed that the AAV-MG variants mediate safe in vivo transgene delivery without inducing microglia immune activation. These AAV variants should facilitate the applications of various genetically-encoded sensors and effectors in studying microglia-related biology and therapeutic interventions.

The analyses presented here are used to support the findings in the original paper published in Nature Methods:  
[Lin et al.]()

## Resources

The sequencing data used in this study can be accessed via GEO with accession number [GSE197743](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE197743).

## List of contents

Assessment of AAV transduction on *in vitro* cultured microglia via RNAseq:

  - [RNAseq vignette - Figure 1](https://ruiyuraywang.github.io/Lin_AAV_Microglia_public/Fig1_rnaseq.html)

Assessment of *in vivo* AAV transduction in microglia by scRNA-seq:

  * Preprocessing
    + [Query Dataset (Smart-seq2) - Supplementary Figure 2](https://ruiyuraywang.github.io/Lin_AAV_Microglia_public/FigS2_scRNAseq_query_QC.html)
    + [Reference Dataset (10X) - Supplementary Figure 2](https://ruiyuraywang.github.io/Lin_AAV_Microglia_public/FigS2_scRNAseq_reference_QC.html)
  * Primary Analyses
    + [Label Transfer, and Comparative Analysis between Query and Reference - Supplementary Figure 2](https://ruiyuraywang.github.io/Lin_AAV_Microglia_public/FigS2_scRNAseq_label_transfer.html)
    + [Descriptive Analysis, Label Transfer, and Odds Ratio - Figure 2](https://ruiyuraywang.github.io/Lin_AAV_Microglia_public/Fig2_main_figs.html)
  * Auxiliary Analysis
    + [Correlation Analysis of Signature Genes and Transgene (mScarlet) - Supplementary Figure 3](https://ruiyuraywang.github.io/Lin_AAV_Microglia_public/FigS3_scRNAseq_corr.html)

## Acknowledgements

This work is supported by
    
    
    