---
title: 'Lin et al.: AAV Microglia'
subtitle: 'scRNA-seq Reference Dataset: 10X'
author: "Ruiyu Ray Wang"
date: "`r Sys.Date()`"
output: html_document
---

```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## Introduction

Characterization of microglia polarization states under targeted AAV transduction using bulk and single cell RNA sequencing.

The study of microglia biology and the development of microglia-based gene therapies are in urgent need of efficient and safe vehicles for microglia transgene delivery. To address this, we developed adeno-associated virus (AAV) variants that mediate efficient in vitro and in vivo microglia transduction via directed evolution of the AAV capsid protein. To assess the effect of AAV transduction on microglia, we carried out bulk RNAseq in primary microglia and found that microglia transduced by AAV remain close to homeostatic state. Furthermore, single-cell RNA sequencing showed that the AAV-MG variants mediate safe in vivo transgene delivery without inducing microglia immune activation. These AAV variants should facilitate the applications of various genetically-encoded sensors and effectors in studying microglia-related biology and therapeutic interventions.

### About this Vignette

This vignette illustrates the steps by which the reference data (10X) in the original publication (Lin et al., 2022) was processed. The input file of this vignette is a filtered `h5` file generated by `CellRanger` (`GSE197743_AAV_MG_filtered_feature_bc_matrix.h5`) . The output is a processed `.h5Seurat` with metadata slot filled.

### Overview of the experiments

Microglia were isolated from adult mice (~8w) using a method based on the "cold-mechanical dissociation" protocol (Bennett et al., 2016)[^1]. The steps involving FACS sorting from the original protocol was omitted to minimize *ex vivo* activation of microglia.

We found that the single cell transcriptome acquired by 10X Chromium (3' Chemistry v3) captured transcripts of the AAV-packaged transgene (mScarlet) at a very low rate (data not published). Therefore we carried out deep single cell sequencing using a modified protocol based on a customized Smart-seq2 protocol (Zhang et al., 2020)[^2] on microglia isolated from mice subjected to stereotaxic AAV injection. The Smart-seq2 based method captured mScarlet transcripts with substantially higher efficiently.

To characterize the microglia transcriptomes under immune activation, we isolated microglia from mice subjected to Saline, Poly(i:c) and LPS treatment (i.p.), and performed scRNA-seq using the 10X Chromium platform.

### Analysis Outline

Upstream analysis pipelines (i.e. sequence alignment, generation of digital expression matrix, etc.):

  * For Smart-seq2, fastq files were processed by a [custom Snakemake workflow](https://github.com/RuiyuRayWang/ScRNAseq_smkpipe_at_Luolab) to generate the expression table (feature-barcode-count matrix).
  * For 10X, fastq files from 3 mice were quantified by CellRanger (v6.0.1) using the `cellranger count` command and then aggregated using the `cellranger aggr` command.

Downstream analysis:

  * Preprocessing and QC (`scater`)
  * Seurat Standard Workflow (`Seurat`)
  * Label Transfer (`Seurat`)

## Pre-processing

### Load libraries and data

Load reuiqred libraries.
```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(SeuratDisk)
  library(scater)
  library(tidyverse)
  library(scales)
})
```

If you haven't done so, download data from GEO.
```{r}
if(!file.exists('../data/GSE197743/GSE197743_AAV_MG_SMARTSEQ2_counts.tsv.gz')){
  library(GEOquery)
  options(timeout = 6000)  ## In case of slow Internet connections
  getGEOSuppFiles(GEO = "GSE197743", makeDirectory = T, baseDir = '../data/')
}
```

Load the data matrix to workspace.
```{r}
data <- Read10X_h5(filename = '../data/GSE197743/GSE197743_AAV_MG_filtered_feature_bc_matrix.h5')
```
Create Seurat object.
```{r}
cells <- CreateSeuratObject(data, project = "AAV_Microglia", names.field = 2, names.delim = "-")
```

Update metadata.
```{r}
Idents(cells) <- "orig.ident"
cells <- RenameIdents(cells, `1` = "MG_S", `2` = "MG_P", `3` = "MG_L")
cells[["group"]] <- Idents(cells)

Idents(cells) <- "orig.ident"
cells <- RenameIdents(cells, `1` = "Saline", `2` = "Poly(i:c)", `3` = "LPS")
cells[["treat"]] <- Idents(cells)

cells[["id"]] <- "Reference"
cells[["tech"]] <- "10x"
```


### Quality Control (QC)

We use the `scater`[^3] package for quality control (QC) of the single cell datasets
```{r}
cells <- DietSeurat(cells)
cells.sce <- as.SingleCellExperiment(cells)
cells.sce <- logNormCounts(cells.sce)
dim(cells.sce)
```

#### Cell level QC

```{r}
per.cell <- perCellQCMetrics(cells.sce,
                             subsets = list(Mito = grep("^mt-", rownames(cells.sce)),  # Mitochondrial genes
                                            Ribo = grep("\\bRp[sl]\\d+[^k]*\\b", rownames(cells.sce))))  # Ribosomal genes
qc.stats <- quickPerCellQC(per.cell, percent_subsets = c("subsets_Mito_percent","subsets_Ribo_percent"))

colData(cells.sce) <- cbind(colData(cells.sce), cbind(per.cell,qc.stats))
colSums(as.matrix(qc.stats))
```
```{r}
cells.filtered <- cells.sce[,!qc.stats$discard]
```

#### Feature level QC

```{r}
keep_feature <- nexprs(cells.filtered, byrow=TRUE) > 0

## Remove mitochondrial and ribosomal genes which does not contribute to dimensionality reduction and clustering
keep_feature[grep("\\bRp[sl]\\d+[^k]*\\b", rownames(cells.sce))] <- FALSE
keep_feature[grep("mt-", rownames(cells.sce))] <- FALSE
cells.filtered <- cells.filtered[keep_feature,]
```

Back to Seurat object.
```{r}
refdata <- as.Seurat(cells.filtered)
refdata[["nCount_originalexp"]] <- NULL; refdata[["nFeature_originalexp"]] <- NULL
refdata <- RenameAssays(refdata, originalexp = "RNA")
refdata$ident <- NULL
```

### Standard Seurat Workflow

```{r}
refdata <- refdata %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA(npcs = 50)

## USE DIM=1:50 FOR UNFILTERED DATA, SKIP JACKSTRAW TO SAVE TIME
# cells <- JackStraw(cells, dims = 50)
# cells <- ScoreJackStraw(cells, dims = 1:50)
# JackStrawPlot(cells, dims = 1:50)
# dims_use <- JS(cells[["pca"]], slot = "overall") %>% as.data.frame %>% filter(Score < 0.05) %>% pull(PC)

dims_use <- 1:50

refdata <- refdata %>%
  FindNeighbors(dims = dims_use) %>%
  FindClusters(resolution = 0.8) %>%
  RunUMAP(dims = dims_use)
```

Take a closer look at our pre-processed data:  
We see that there are some major clusters in the middle-right of the umap plot. These are microglia single cell populations. The rest of the clusters represent non-microglia populations or droplets contaminated by cell debris. They were captured because density gradient centrifugation by Percoll can never perfectly purify microglia from brains subjected to Dounce homogenization. Nevertheless, the microglia clusters look good as they exhibit robust `Hexb` expression. We can easily isolate them via *in silico* procedures, as is shown in subsequent code chunks in this vignettes below.
```{r fig.width=10, fig.height=4}
FeaturePlot(refdata, features = "Hexb") | FeaturePlot(refdata, features = "subsets_Mito_percent")
```

Some microglia single cells show high mitochondrial levels, purge them.
```{r}
cells_sub <- subset(refdata, subset = subsets_Mito_percent < 3)
```

To achieve *in silico* purification of microglia populations, we use the `CellSelector()` utility provided by `Seurat` package, as illustrated in the code chunk below. This procedure requires manual selection of cells and is difficult to replicate programmatically.  
To help you reproduce the result, we have saved the identity of the selected cells in `../miscs/mg_sub.rda` in this repo. You can load it directly into workspace to subset out the micorglia population and proceed to downstream workflows.
```{r}
## NOT RUN
# Idents(cells_sub) <- "contaminants"
# cells_sub <- CellSelector(FeaturePlot(cells_sub, features = "Hexb"), object = cells_sub, ident = "microglia")  ## Run repeatedly, select desired clusters
# mg_sub <- WhichCells(cells_sub, idents = "microglia")
# save(mg_sub, file = "../miscs/mg_sub.rda")
```
```{r}
load(file = '../miscs/mg_sub.rda')
cells_sub <- subset(cells_sub, cells = mg_sub)
```

Re-run Seurat Standard workflow on subsetted data.
```{r}
cells_sub <- cells_sub %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA(npcs = 45) %>%
  FindNeighbors(dims = 1:45) %>%
  FindClusters() %>%
  RunUMAP(dims = 1:45)
```

Take a glance at the umap reduction plot.
```{r fig.width=6, fig.height=4}
DimPlot(cells_sub, group.by = "treat", cols = hue_pal()(3)[c(3,2,1)])
```
Depending on the use of CPU nodes or versions of the `umap` package, the UMAP coordinates computed on different machines may not entirely match what was shown in the original publication. This is a known [issue](https://github.com/lmcinnes/umap/issues/525) of the UMAP algorithm. Nevertheless, the overall structure of the embedding should be consistent, i.e. query cells from LPS and AAV groups reside in different clusters. Therefore we conclude that the shifts in UMAP coordinates does not affect our biological interpretation.

### Update Metadata

The metadata slot is updated based on the experiment design.  
But note that in recent years the microglia community is gradually quitting the use of `Reactive` and `Homeostatic` nomenclature.
```{r}
Idents(cells_sub) <- "orig.ident"
cells_sub <- RenameIdents(cells_sub, `1` = "Homeostatic", `2` = "Reactive", `3` = "Reactive")
cells_sub[["state"]] <- Idents(cells_sub)
```

## Save Results

Backup the processed Seurat object using `h5` format.
```{r}
## Avoid using factor while saving to other formats
cells_sub$group <- as.character(cells_sub$group)
cells_sub$treat <- as.character(cells_sub$treat)
cells_sub$state <- as.character(cells_sub$state)
cells_sub$orig.ident <- as.character(cells_sub$orig.ident)
cells_sub$RNA_snn_res.0.8 <- as.integer(as.character(cells_sub$RNA_snn_res.0.8))
cells_sub$seurat_clusters <- as.integer(as.character(cells_sub$seurat_clusters))

if(!dir.exists('../data/AAV_MG_scRNAseq')){dir.create('../data/AAV_MG_scRNAseq')}
SaveH5Seurat(cells_sub, filename = '../data/AAV_MG_scRNAseq/refdata_clean.h5Seurat', overwrite = T, verbose = F)
Convert('../data/AAV_MG_scRNAseq/refdata_clean.h5Seurat', dest = "h5ad", overwrite = T, verbose = F)
```

### Session info

```{r}
sessionInfo()
```

[^1]: Bennett, M. L. et al. New tools for studying microglia in the mouse and human CNS. Proc Natl Acad Sci USA 113, E1738 (2016).
[^2]: Zhang, S. et al. Single-cell transcriptomics identifies divergent developmental lineage trajectories during human pituitary development. Nat Commun 11, 5275 (2020).
[^3]: McCarthy, D. J., Campbell, K. R., Lun, A. T. L. & Wills, Q. F. Scater: pre-processing, quality control, normalization and visualization of single-cell RNA-seq data in R. Bioinformatics btw777 (2017) doi:10.1093/bioinformatics/btw777.
