---
title: 'Lin et al.: AAV Microglia'
subtitle: 'scRNA-seq Query Dataset: Smartseq2'
author: "Ruiyu Ray Wang"
date: "`r Sys.Date()`"
output: html_document
---

```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## Introduction

Characterization of microglia polarization states under targeted AAV transduction using bulk and single cell RNA sequencing.

The study of microglia biology and the development of microglia-based gene therapies are in urgent need of efficient and safe vehicles for microglia transgene delivery. To address this, we developed adeno-associated virus (AAV) variants that mediate efficient in vitro and in vivo microglia transduction via directed evolution of the AAV capsid protein. To assess the effect of AAV transduction on microglia, we carried out bulk RNAseq in primary microglia and found that microglia transduced by AAV remain close to homeostatic state. Furthermore, single-cell RNA sequencing showed that the AAV-MG variants mediate safe in vivo transgene delivery without inducing microglia immune activation. These AAV variants should facilitate the applications of various genetically-encoded sensors and effectors in studying microglia-related biology and therapeutic interventions.

### About this Vignette

This vignette illustrates the steps by which the query data (Smartseq2) in the original publication (Lin et al., 2022) was processed. The input file is a feature-barcode-count table (`GSE197743_AAV_MG_SMARTSEQ2_counts.tsv.gz`) produced by the Snakemake pipeline [`ScRNAseq_smkpipe_at_Luolab`](https://github.com/RuiyuRayWang/ScRNAseq_smkpipe_at_Luolab). The output of this vignette is a processed `.h5Seurat` with metadata slot filled.s

### Overview of the experiment

Microglia were isolated from adult mice (~8w) using a method based on the "cold-mechanical dissociation" protocol (Bennett et al., 2016)[^1].

We found that the single cell transcriptome acquired by 10X Chromium (3' Chemistry v3) captured transcripts of the AAV-packaged transgene (mScarlet) at a very low rate (data not published). Therefore we carried out deep single cell sequencing using a modified protocol based on a customized Smart-seq2 protocol (Zhang et al., 2020)[^2] on microglia isolated from mice subjected to stereotaxic AAV injection. The Smart-seq2 based method captured mScarlet transcripts with substantially higher efficiently.

To characterize the microglia transcriptomes under immune activation, we isolated microglia from mice subjected to Saline, Poly(i:c) and LPS treatment (i.p.), and performed scRNA-seq using the 10X Chromium platform.

### Analysis Outline

Upstream analysis pipelines (i.e. sequence alignment, generation of digital expression matrix, etc.):

  * For Smart-seq2, fastq files were processed by a [custom Snakemake workflow](https://github.com/RuiyuRayWang/ScRNAseq_smkpipe_at_Luolab) to generate the expression table (feature-barcode-count matrix).
  * For 10X, fastq files from 3 mice were quantified by CellRanger (v6.0.1) using the `cellranger count` command and then aggregated using the `cellranger aggr` command.

Downstream analysis:

  * Preprocessing and QC (`scater`)
  * Seurat Standard Workflow (`Seurat`)
  * Label Transfer (`Seurat`)

## Pre-processing

### Load libraries and data

Load reuiqred libraries.
```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(SeuratDisk)
  library(scater)
  library(tidyverse)
})
```

If you haven't done so, download data from GEO.
```{r}
if(!file.exists('../data/GSE197743/GSE197743_AAV_MG_SMARTSEQ2_counts.tsv.gz')){
  library(GEOquery)
  options(timeout = 6000)  ## In case of slow Internet connections
  getGEOSuppFiles(GEO = "GSE197743", makeDirectory = T, baseDir = '../data/')
}
```

Load the data matrix to workspace.
```{r}
data_long <- read.table(file = '../data/GSE197743/GSE197743_AAV_MG_SMARTSEQ2_counts.tsv.gz', sep = "\t", header = TRUE)  # Raw table from ScRNAseq_smkpipe_at_Luolab
data <- pivot_wider(data_long, names_from = "cell", values_from = "count", values_fill = 0)  # Expression matrix
```

Rename Ensembl ID to Gene symbol.
```{r}
# `../miscs/MM.GRCm38.102.annotation.tab` is generated by running `../misc/awk_extract_gtf.sh` on the annotation file (Mus_musculus.GRCm38.102.chr.mScarlet_WPRE.gtf)
anno <- read.table(file = '../miscs/MM.GRCm38.102.annotation.tab', sep = "\t", col.names = c("ensembl_id", "symbol"))
```

```{r}
# Custom function to rename rows from dataframe
renameRows <- function(df, anno){
  stopifnot(is.data.frame(df))
  stopifnot("ensembl_id" %in% colnames(anno) | "symbol" %in% colnames(anno))
  
  if(any(!rownames(df) %in% anno$ensembl_id)){
    df <- dplyr::slice(df, -which(!rownames(df) %in% anno$ensembl_id))  # df may contain rows not in the annotation table. Remove these rows.
  }
  rn <- rownames(df)
  rn <- anno[match(rn,anno$ensembl_id),"symbol"]
  dup <- rn[duplicated(rn)]
  print(paste0("The following genes names are duplicated: ", paste0(dup, collapse = ", "), ". Making unique."))
  for (d in dup){
    count = 0
    for (i in 1:length(rn)){
      if (rn[i] == d) {
        count = count+1
        if (count>1){
          rn[i] = paste0(rn[i],"-",count)
        }
      }
    }
  }
  rownames(df) <- rn
  return(df)
}
```
```{r}
data <- data %>% column_to_rownames(var = "gene") %>% as.data.frame()
data <- renameRows(data, anno)
```

Transgene (mScarlet) should be moved to metadata slot to avoid its impact on downstream analyses.
```{r}
ms <- data.frame(t(data["mScarlet",])); colnames(ms) <- "mScarlet"
data <- data[-which(rownames(data) == "mScarlet"),]
```

Create Seurat object.
```{r}
cells <- CreateSeuratObject(data, project = "AAV_Microglia", names.field = 2, names.delim = "_", meta.data = ms)
cells$batch <- cells$orig.ident
cells$id <- "Query"
cells$tech <- "Smart-seq2"
```

### Quality Control (QC)

We use the `scater`[^3] package for quality control (QC) of the single cell datasets
```{r}
# cells <- DietSeurat(cells)
cells.sce <- as.SingleCellExperiment(cells)
cells.sce <- logNormCounts(cells.sce)
dim(cells.sce)
```

#### Cell level QC

```{r}
per.cell <- perCellQCMetrics(cells.sce,
                             subsets = list(Mito = grep("^mt-", rownames(cells.sce)),  # Mitochondrial genes
                                            Ribo = grep("\\bRp[sl]\\d+[^k]*\\b", rownames(cells.sce))))  # Ribosomal genes
colData(cells.sce) <- cbind(colData(cells.sce), per.cell)

qc.stats <- quickPerCellQC(per.cell, percent_subsets = c("subsets_Mito_percent","subsets_Ribo_percent"))
colData(cells.sce) <- cbind(colData(cells.sce), qc.stats)
colSums(as.matrix(qc.stats))
```
```{r}
cells.filtered <- cells.sce[,!qc.stats$discard]
```

#### Feature level QC

```{r}
keep_feature <- nexprs(cells.filtered, byrow=TRUE) > 0

## Remove mitochondrial and ribosomal genes which does not contribute to dimensionality reduction and clustering
keep_feature[grep("\\bRp[sl]\\d+[^k]*\\b", rownames(cells.sce))] <- FALSE
keep_feature[grep("mt-", rownames(cells.sce))] <- FALSE
cells.filtered <- cells.filtered[keep_feature,]
```

Back to Seurat object.
```{r}
querydata <- as.Seurat(cells.filtered)
querydata[["nCount_originalexp"]] <- NULL; querydata[["nFeature_originalexp"]] <- NULL
querydata <- RenameAssays(querydata, originalexp = "RNA")
querydata$ident <- NULL
```

### Standard Seurat Workflow

```{r}
dims_use <- 1:25

querydata <- querydata %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA(npcs = 50) %>%
  FindNeighbors(dims = dims_use) %>%
  FindClusters(resolution = 0.8) %>%
  RunTSNE(dims = dims_use) %>%
  RunUMAP(dims = dims_use)
```

Remove cells NOT expressing the canonical microglia marker 'Hexb'. These are putatively low quality cells.
```{r fig.width=12, fig.height=5}
FeaturePlot(querydata, features = "Hexb") | DimPlot(querydata, cells.highlight = WhichCells(querydata, expression = Hexb < 2))
```
Re-do Seurat workflow on subsetted data.
```{r}
querydata <- subset(querydata, subset = Hexb > 2)
querydata <- querydata %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA(npcs = 50) %>%
  FindNeighbors(dims = dims_use) %>%
  FindClusters(resolution = 0.2) %>%
  RunTSNE(dims = dims_use) %>%
  RunUMAP(dims = dims_use)
```

In the figure below, we can clearly see that microglia from LPS treated mice were primed and adopted a distinct transcriptomic state. Later, we'll use [`Label Transfer`](https://satijalab.org/seurat/articles/integration_mapping.html) to analyze them more closely.
```{r fig.width=12, fig.height=5}
DimPlot(querydata) | DimPlot(querydata, group.by = "orig.ident")
```

### Update Metadata

The metadata slot is updated based on the experiment design.
```{r}
Idents(querydata) <- "orig.ident"
querydata <- SetIdent(querydata, cells = WhichCells(querydata, expression = mScarlet > 0), value = "Transduced")
querydata <- SetIdent(querydata, cells = WhichCells(querydata, expression = mScarlet == 0), value = "Non-transduced")
querydata[["transduction"]] <- Idents(querydata)

Idents(querydata) <- "orig.ident"
querydata <- RenameIdents(querydata, 
                          `yt-8-5` = "AAV-MG1.1", `yt-8-5-2` = "AAV-MG1.1", `yt-8-5-3` = "AAV-MG1.1",
                          `yt-8-5-4` = "AAV-MG1.1", `yt-8-5-6` = "AAV-MG1.1", `yt-8-5-5` = "AAV-MG1.1",
                          `yt-8-6-1` = "AAV-MG1.2", `yt-8-6-2` = "AAV-MG1.2", `yt-8-6-3` = "AAV-MG1.2", 
                          `yt-8-6-4` = "AAV-MG1.2", `yt-8-6-5` = "AAV-MG1.2", `yt-8-6-6` = "AAV-MG1.2",
                          `yt-mgl-L1` = "LPS", `yt-mgl-L2` = "LPS", `yt-mgl-L3` = "LPS")
querydata[["treatment"]] <- Idents(querydata)
querydata[["treatment"]] <- factor(querydata$treatment, 
                                   levels = c("AAV-MG1.2", "AAV-MG1.1","LPS"))

Idents(querydata) <- "treatment"
querydata <- RenameIdents(querydata, `AAV-MG1.1` = "AAV", `AAV-MG1.2` = "AAV")
querydata[["group"]] <- Idents(querydata)

Idents(querydata) <- "treatment"
querydata <- SetIdent(querydata, 
                      cells = WhichCells(querydata,
                                         expression = transduction == "Non-transduced" & 
                                           treatment %in% c("AAV-MG1.1","AAV-MG1.2")), 
                      value = "Non-transduced")
querydata[["transduce_serotype"]] <- Idents(querydata)
querydata[["transduce_serotype"]] <- factor(querydata$transduce_serotype, 
                                            levels = c("Non-transduced","AAV-MG1.2", "AAV-MG1.1","LPS"))

querydata[["mScarlet_norm"]] <- log1p(querydata[["mScarlet"]]/querydata[["nCount_RNA"]]*1e4)
```

## Save Results

Backup the processed Seurat object using `h5` format.
```{r}
## Avoid using factor while saving to other formats
querydata$orig.ident <- as.character(querydata$orig.ident)
querydata$batch <- as.character(querydata$batch)
querydata$transduction <- as.character(querydata$transduction)
querydata$treatment <- as.character(querydata$treatment)
querydata$group <- as.character(querydata$group)
querydata$transduce_serotype <- as.character(querydata$transduce_serotype)
querydata$seurat_clusters <- as.integer(as.character(querydata$seurat_clusters))
SaveH5Seurat(querydata, filename = '../data/AAV_MG_scRNAseq/cells_query.h5Seurat', overwrite = T)
Convert('../data/AAV_MG_scRNAseq/cells_query.h5Seurat', dest = "h5ad", overwrite = T)
```

### Session info

```{r}
sessionInfo()
```

[^1]: Bennett, M. L. et al. New tools for studying microglia in the mouse and human CNS. Proc Natl Acad Sci USA 113, E1738 (2016).
[^2]: Zhang, S. et al. Single-cell transcriptomics identifies divergent developmental lineage trajectories during human pituitary development. Nat Commun 11, 5275 (2020).
[^3]: McCarthy, D. J., Campbell, K. R., Lun, A. T. L. & Wills, Q. F. Scater: pre-processing, quality control, normalization and visualization of single-cell RNA-seq data in R. Bioinformatics btw777 (2017) doi:10.1093/bioinformatics/btw777.
