---
title: "Supplementary Figure 1: AAV priming effect on Microglia"
author: "Ruiyu Ray Wang"
date: "`r Sys.Date()`"
output: html_document
---

### Overview of the experiment

Analysis pipeline:
  
  - Salmon (Pseudo-alignment and quantification)
  - DESeq (differential expression analysis)

<br>

```{r include=FALSE}
rm(list = ls(all.names = TRUE))
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

# Initialization

```{r}
suppressPackageStartupMessages({
  library(tximeta)
  library(SummarizedExperiment)
  library(DESeq2)
  library(ggplot2)
  library(GGally)
  library(tidyverse)
  library(genefilter)
  library(RColorBrewer)
  library(pheatmap)
})
```

```{r}
suppressMessages(
  extrafont::loadfonts(device="postscript")
)
```

Load design table.
```{r}
coldata <- read.csv('../misc/RNAseq_sample_table.csv', stringsAsFactors=FALSE) %>% 
  mutate(
    files = file.path("../data/AAV_MG_RNAseq/",lib,"quants",SampleName,"quant.sf"),
    names = SampleName
    ) %>%
  column_to_rownames("SampleName")

# To avoid warnings at creating dds, convert integer to factor, recode levels to only contain "_" and "."
coldata <- coldata %>%
  mutate(
    stimuli = recode(stimuli, `LPS_200ng/mL` = "LPS_200ng_mL", `IL4_20ng/mL` = "IL4_20ng_mL"),
    virus = recode(virus, 
                   `cap8_mScarlet, moi=50,000` = "cap8_mSc_moi_50000", 
                   `AAV-cap18-msc, moi=50,000` = "cap18_mSc_moi_50000",
                   `cap18_mScarlet, moi=50,000` = "cap18_mSc_moi_50000")
  ) %>%
  mutate(
    batch = factor(batch),
    stimuli = factor(stimuli),
    virus = factor(virus)
    )
coldata
```

```{r include=FALSE}
coldata <- coldata[c(
  # "3-O","3-P","1-G","1-H","1-I","1-J","2-I","2-J","2-K","2-L",
  "yt0616-1","yt0616-2","yt0616-3",
  # "yt0616-4","yt0616-5","yt0616-6",
  # "yt0616-7","yt0616-8","yt0616-9",
  # "yt0616-13","yt0616-14","yt0616-15",
  # "yt0616-16","yt0616-17","yt0616-18",
  "YT-1111-01","YT-1111-02","YT-1111-03",
  "YT-1111-04","YT-1111-05","YT-1111-06",
  "YT-1111-07","YT-1111-08","YT-1111-09",
  # "YT-1111-10","YT-1111-11",
  "YT-1111-12"),]  ## Select sample
writeLines("Confirm that file search paths are parsed correctly:")
file.exists(coldata$files)
```

```{r include=FALSE}
# tximeta
se <- tximeta(coldata)
dim(se)
head(rownames(se))
colnames(se)

# summarize to gene
gse <- summarizeToGene(se)

# peek
gse
assayNames(gse)
head(assay(gse), 3)

colSums(assay(gse))
rowRanges(gse)
seqinfo(rowRanges(gse))
colData(gse)

round( colSums(assay(gse)) / 1e6, 1 )  # Millions of reads for each library
```

```{r}
# dds obj
dds <- DESeqDataSet(gse, design = ~ stimuli + virus)
```

Pre-filtering.
```{r include=FALSE}
nrow(dds)

# keep <- rowSums(counts(dds)) > 1
keep <- rowSums(counts(dds) >= 10) >= 3  # at least 3 samples with a count of 10 or higher
dds <- dds[keep,]
nrow(dds)

# # Using `genefilter` package, extract genes with the highest variance (upper 75 percentile).
# rv = rowVars(assay(dds, i = "abundance"))
# q75 = quantile(rowVars(assay(dds, i = "abundance")), .85)
# hvg_keep = rv > q75
# dds <- dds[hvg_keep,]
# nrow(dds)
```

```{r include=FALSE}
# IMPORTANT: Reset reference levels
# If this step is omitted, DESeq will perform comparisons based on the alphabetical order of the levels, which will lead to error
dds$stimuli <- factor(dds$stimuli, levels = c("blank", "IL4_20ng_mL","LPS_200ng_mL"))
dds$virus <- factor(dds$virus, levels = c("untrt", 
                                          # "cap8_mSc_moi_50000", 
                                          "cap18_mSc_moi_50000"))
```

#### Exploratory analysis: Sample Distance & PCA

Variance stablizing transformation.
```{r include=FALSE}
vsd <- vst(dds, blind = FALSE)
head(assay(vsd), 3)
colData(vsd)
```

#### Batch correction.

```{r}
mat <- assay(vsd)
mat <- limma::removeBatchEffect(mat, vsd$batch)
assay(vsd) <- mat
```

#### Sample distance

```{r include=FALSE}
sampleDists <- dist(t(assay(vsd)))
```

```{r fig.height=7}
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste0(sub("_.*","",vsd$stimuli), " - ", sub("_.*","",vsd$virus), " - batch", vsd$batch)
# colnames(sampleDistMatrix_bc) <- colnames(sampleDistMatrix_bc)
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
out <- pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

Obtain Hierachical clustering result shown above, use later.
```{r}
clusters <- cutree(out$tree_col, k = 2)
vsd$clusters <- as.factor(clusters)
dds$clusters <- as.factor(clusters)
```

#### Supplementary Figure 1a: PCA

```{r fig.width=9, fig.height=5}
pcaData <- plotPCA(vsd, intgroup = c("names","clusters","stimuli","virus"), returnData = TRUE)
percentVar <- round(100* attr(pcaData, "percentVar"))
## Demo
ggplot(pcaData, aes(x = PC1, y = PC2)) + 
  geom_point(aes(color = clusters, shape = virus, fill = stimuli), size = 2, stroke = .6) + 
  scale_shape_manual("Virus", values = c(21,22,24), labels = c("Untransfected", "AAV-cMG")) +
  scale_fill_manual(values=c("white",brewer.pal(n = 12, name = "Paired")[c(5,3)])) +
  guides(color = "none",
         fill = guide_legend(override.aes = list(shape = 21))) +  # Hack: github.com/tidyverse/ggplot2/issues/2322
  stat_ellipse(data=pcaData, mapping=aes(color = clusters), linetype = 2) +
  # xlim(-30,60) + ylim(-28,28) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) + 
  ylab(paste0("PC2: ", percentVar[2], "% variance")) + 
  coord_fixed() +
  theme_bw()

## Actual plot
ggplot(pcaData, aes(x = PC1, y = PC2)) + 
  geom_point(aes(color = clusters, shape = virus, fill = stimuli), size = 2, stroke = .6) + 
  scale_shape_manual("Virus", values = c(21,22,24), labels = c("Untransfected", "AAV-cMG")) +
  scale_fill_manual(values=c("white",brewer.pal(n = 12, name = "Paired")[c(5,3)])) +
  guides(color = "none",
         fill = guide_legend(override.aes = list(shape = 21))) +  # Hack: github.com/tidyverse/ggplot2/issues/2322
  stat_ellipse(data=pcaData, mapping=aes(color = clusters), linetype = 2) +
  # xlim(-30,60) + ylim(-28,28) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) + 
  ylab(paste0("PC2: ", percentVar[2], "% variance")) + 
  coord_fixed() +
  theme_bw() +
  theme(
    legend.position = "none"
  )
```
```{r}
ggsave(filename = "PCA_priming_effect.eps", 
       device = "eps",
       path = "../outs/Figsupp1/",
       width = 4, height = 3,
       dpi = 300,
       family = "Arial"
       )
```

### Priming effect

```{r include=FALSE}
coldata <- coldata[c(
  "YT-1111-01","YT-1111-02","YT-1111-03",
  "YT-1111-04","YT-1111-05","YT-1111-06",
  "YT-1111-07","YT-1111-08","YT-1111-09",
  "YT-1111-12"),]  ## Batch 3
writeLines("Confirm that file search paths are parsed correctly:")
file.exists(coldata$files)
```

```{r include=FALSE}
# tximeta
se <- tximeta(coldata)
dim(se)
head(rownames(se))
colnames(se)

# summarize to gene
gse <- summarizeToGene(se)

# peek
gse
assayNames(gse)
head(assay(gse), 3)

colSums(assay(gse))
rowRanges(gse)
seqinfo(rowRanges(gse))
colData(gse)

round( colSums(assay(gse)) / 1e6, 1 )  # Millions of reads for each library
```

```{r}
# dds obj
dds <- DESeqDataSet(gse, design = ~ stimuli + virus)
```

Pre-filtering.
```{r include=FALSE}
nrow(dds)

# keep <- rowSums(counts(dds)) > 1
keep <- rowSums(counts(dds) >= 10) >= 3  # at least 3 samples with a count of 10 or higher
dds <- dds[keep,]
nrow(dds)
```

#### Variance stablizing transformation

```{r include=FALSE}
vsd <- vst(dds, blind = FALSE)
head(assay(vsd), 3)
colData(vsd)
```

#### Supplementary Figure 1b: Sample distance

```{r include=FALSE}
sampleDists <- dist(t(assay(vsd)))
```

```{r fig.height=7}
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste0(sub("_.*","",vsd$stimuli), " - ", 
                                     recode(vsd$virus, "cap18_mSc_moi_50000" = "AAV-cMG", "untrt" = "Untransduced"))
# colnames(sampleDistMatrix_bc) <- colnames(sampleDistMatrix_bc)
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
out <- pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         labels_col = rep("",10),
         col = colors)
```
