---
title: "Supplementary Figure 3: Query Data Transduction and Marker Gene Expression"
author: "Ruiyu Ray Wang"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
rm(list = ls(all.names = TRUE))
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

# Initialization

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(SeuratDisk)
  library(tidyverse)
  library(ggplot2)
  library(ggpubr)
  library(patchwork)
  library(RColorBrewer)
})
```

```{r}
suppressMessages(
  extrafont::loadfonts(device="postscript")
)
```

```{r}
querydata <- LoadH5Seurat('../data/AAV_MG_scRNAseq/cells_query.h5Seurat')
query_transferred <- LoadH5Seurat('../data/AAV_MG_scRNAseq/cells_query_transferred.h5Seurat')
```

```{r}
queryaav_sub <- subset(querydata, subset = group == "AAV")
query_transduced <- subset(queryaav_sub, subset = transduction == "Transduced")
```

# Making Plots

### AAV treated group

#### Supplementary Figure 3a

```{r}
features_hom = c("P2ry12","P2ry13","Cx3cr1","Csf1r","Lrp1","Ptpre","Lipa","Ets1")
features_hom_lab = c("Sall1", "Sirpa")

vplots <- list()
for(feature in features_hom){
  vplots[[feature]] <- FetchData(queryaav_sub, vars = c("treatment", "transduction", feature, "mScarlet_norm")) %>%
    ggplot(aes_string(x = "treatment", y = feature)) + 
    geom_jitter(aes_string(color = "mScarlet_norm"), size = 2, alpha = 1, width = 0.20) + 
    geom_violin(aes_string(fill = "treatment"), alpha = 0.3, width = 0.6) +
    scale_color_distiller('normalized\nmScarlet\nexpression', palette = "Reds", direction = 1) +
    scale_fill_manual(values = brewer.pal(n = 12, name = "Paired")[c(11,7)]) +
    guides(fill = "none") +
    theme_pubr() +
    theme(
      axis.title.x = element_blank(),
      axis.text.x = element_blank(),
      axis.title.y = element_text(size = 22, family = "Arial"),
      axis.text.y = element_text(size = 18, family = "Arial"),
      legend.position = "none",
      legend.title = element_text(size = 17, family = "Arial"),
      legend.text = element_text(size = 16, family = "Arial")
    )
}

for(feature in features_hom_lab){
  vplots[[feature]] <- FetchData(queryaav_sub, vars = c("treatment", "transduction", feature, "mScarlet_norm")) %>%
    ggplot(aes_string(x = "treatment", y = feature)) + 
    geom_jitter(aes_string(color = "mScarlet_norm"), size = 2, alpha = 1, width = 0.20) + 
    geom_violin(aes_string(fill = "treatment"), alpha = 0.3, width = 0.6) +
    scale_color_distiller('normalized\nmScarlet\nexpression', palette = "Reds", direction = 1) +
    scale_fill_manual(values = brewer.pal(n = 12, name = "Paired")[c(11,7)]) +
    guides(fill = "none") +
    theme_pubr() +
    theme(
      axis.title.x = element_blank(),
      axis.text.x = element_text(size = 18, family = "Arial"),
      axis.title.y = element_text(size = 22, family = "Arial"),
      axis.text.y = element_text(size = 18, family = "Arial"),
      legend.position = "none",
      legend.title = element_text(size = 17, family = "Arial"),
      legend.text = element_text(size = 16, family = "Arial"),
    )
}

cairo_ps(filename = 'outs/Figsupp3/query_hom_vlns.eps', width = 12, height = 9)
ggarrange(vplots[["P2ry12"]], vplots[["P2ry13"]], vplots[["Cx3cr1"]], vplots[["Csf1r"]], vplots[["Sall1"]], vplots[["Sirpa"]], ncol=2, nrow=3, common.legend = TRUE, legend="right")
dev.off()
```

#### Supplementary Figure 3b

These are not required by the reviewers but I think they do better at illustrating the points.
```{r}
scplot <- list()
for(feature in c(features_hom, features_hom_lab)){
  scplot[[feature]] <- FetchData(query_transduced, vars = c(feature, "mScarlet_norm")) %>%
    ggscatter(
      x = "mScarlet_norm",
      y = feature,
      xlab = "Normalized mScarlet expression",
      add = c("reg.line"),
      add.params = list(color = "blue", linetype = "dashed")
    ) +
    stat_cor(
      label.x.npc = 0.70,
      label.y.npc = 0.8,
      label.sep = "\n",
      color = "blue",
      cor.coef.name = "rho",
      method = "spearman",
      size = 5
    ) +
    theme(
      axis.title.x = element_text(size = 16),
      axis.text.x = element_text(size = 16),
      axis.title.y = element_text(size = 18),
      axis.text.y = element_text(size = 16),
    )
}

(scplot[["P2ry12"]] | scplot[["P2ry13"]]) / (scplot[["Cx3cr1"]] | scplot[["Csf1r"]]) / (scplot[["Sall1"]] | scplot[["Sirpa"]])
ggsave(filename = 'query_hom_scatter.eps',
       device = "eps",
       width = 8, height = 9,
       path = '../outs/Figsupp3/',
       family = "Arial"
       )
```

#### Supplementary Figure 3c

```{r}
features_rea = c("Stat3","Gbp7","Ptgds","Irf7","Irf9","Saa3","Tlr2","Mt1","Mt2","Cd74","Lcn2","Tnf","Il1b","Cst7","Cxcl10",
                 "Ifitm3","Spp1")
features_rea_lab = c("Ccl5","Ccl2")

vplots <- list()
for(feature in features_rea){
  vplots[[feature]] <- FetchData(queryaav_sub, vars = c("treatment", "transduction", feature, "mScarlet_norm")) %>%
    ggplot(aes_string(x = "treatment", y = feature)) + 
    geom_jitter(aes_string(color = "mScarlet_norm"), size = 2, alpha = 1, width = 0.20) + 
    geom_violin(aes_string(fill = "treatment"), alpha = 0.3, width = 0.6) +
    scale_color_distiller('normalized\nmScarlet\nexpression', palette = "Reds", direction = 1) +
    scale_fill_manual(values = brewer.pal(n = 12, name = "Paired")[c(11,7)]) +
    guides(fill = "none") +
    theme_pubr() +
    theme(
      axis.title.x = element_blank(),
      axis.text.x = element_blank(),
      axis.title.y = element_text(size = 22, family = "Arial"),
      axis.text.y = element_text(size = 18, family = "Arial"),
      legend.position = "none",
      # legend.key.size = unit(.4, 'cm'),
      # legend.key.height = unit(.4, 'cm'),
      # legend.key.width = unit(.4, 'cm'),
      legend.title = element_text(size = 17, family = "Arial"),
      legend.text = element_text(size = 16, family = "Arial")
    )
}

for(feature in features_rea_lab){
  vplots[[feature]] <- FetchData(queryaav_sub, vars = c("treatment", "transduction", feature, "mScarlet_norm")) %>%
    ggplot(aes_string(x = "treatment", y = feature)) + 
    geom_jitter(aes_string(color = "mScarlet_norm"), size = 2, alpha = 1, width = 0.20) + 
    geom_violin(aes_string(fill = "treatment"), alpha = 0.3, width = 0.6) +
    scale_color_distiller('normalized\nmScarlet\nexpression', palette = "Reds", direction = 1) +
    scale_fill_manual(values = brewer.pal(n = 12, name = "Paired")[c(11,7)]) +
    guides(fill = "none") +
    theme_pubr() +
    theme(
      axis.title.x = element_blank(),
      axis.text.x = element_text(size = 18, family = "Arial"),
      axis.title.y = element_text(size = 22, family = "Arial"),
      axis.text.y = element_text(size = 18, family = "Arial"),
      legend.position = "none",
      # legend.key.size = unit(.4, 'cm'),
      # legend.key.height = unit(.4, 'cm'),
      # legend.key.width = unit(.4, 'cm'),
      legend.title = element_text(size = 17, family = "Arial"),
      legend.text = element_text(size = 16, family = "Arial"),
    )
}

cairo_ps(filename = 'outs/Figsupp3/query_rea_vlns.eps', width = 12, height = 12)
ggarrange(vplots[["Tnf"]], vplots[["Il1b"]], vplots[["Cd74"]], vplots[["Cst7"]], vplots[["Spp1"]], vplots[["Irf9"]], vplots[["Ccl2"]], vplots[["Ccl5"]], ncol=2, nrow=4, common.legend = TRUE, legend="right")
dev.off()
```

#### Supplementary Figure 3d

```{r}
scplot <- list()
for(feature in c(features_rea, features_rea_lab)){
  scplot[[feature]] <- FetchData(query_transduced, vars = c(feature, "mScarlet_norm")) %>%
    ggscatter(
      x = "mScarlet_norm",
      y = feature,
      xlab = "Normalized mScarlet expression",
      add = c("reg.line"),
      add.params = list(color = "blue", linetype = "dashed")
    ) +
    stat_cor(
      label.x.npc = 0.70,
      label.y.npc = 0.8,
      label.sep = "\n",
      color = "blue",
      cor.coef.name = "rho",
      method = "spearman",
      size = 5
    ) +
    theme(
      axis.title.x = element_text(size = 16),
      axis.text.x = element_text(size = 16),
      axis.title.y = element_text(size = 18),
      axis.text.y = element_text(size = 16),
    )
}

(scplot[["Tnf"]] | scplot[["Il1b"]]) / (scplot[["Cd74"]] | scplot[["Cst7"]]) / (scplot[["Spp1"]] | scplot[["Irf9"]]) / (scplot[["Ccl2"]] | scplot[["Ccl5"]])
ggsave(filename = 'query_rea_scatter.eps',
       device = "eps",
       width = 8, height = 12,
       path = '../outs/Figsupp3/',
       family = "Arial"
       )
```
