---
title: "Supplementary Figure 2: Label Transfer Query/Reference Data"
author: "Ruiyu Ray Wang"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
rm(list = ls(all.names = TRUE))
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

# Initialization

```{r}
suppressPackageStartupMessages({
  library(scater)
  library(Seurat)
  library(SeuratDisk)
  library(tidyverse)
  library(ggplot2)
  library(patchwork)
  library(RColorBrewer)
})
```

```{r}
suppressMessages(
  extrafont::loadfonts(device="postscript")
)
```

# Initialization

```{r}
querydata <- LoadH5Seurat('../data/AAV_MG_scRNAseq/cells_query.h5Seurat')
query_transferred <- LoadH5Seurat('../data/AAV_MG_scRNAseq/cells_query_transferred.h5Seurat')
refdata <- LoadH5Seurat('../data/AAV_MG_scRNAseq/refdata_clean.h5Seurat')
refquery <- LoadH5Seurat('../data/AAV_MG_scRNAseq/refquery.h5Seurat')
```

# Making Plots

## Quality control (QC)

### Reference data (10x)

```{r}
data <- Read10X(data.dir = '../data/AAV_MG_scRNAseq/MGREF_AGGR/outs/count/filtered_feature_bc_matrix/')
refdata_raw <- CreateSeuratObject(data, project = "MG_REF", names.field = 2, names.delim = "-")

# Setup metadata
Idents(refdata_raw) <- "orig.ident"
refdata_raw <- RenameIdents(refdata_raw, `1` = "MG_S", `2` = "MG_P", `3` = "MG_L")
refdata_raw[["group"]] <- Idents(refdata_raw)

Idents(refdata_raw) <- "orig.ident"
refdata_raw <- RenameIdents(refdata_raw, `1` = "Saline", `2` = "Poly(i:c)", `3` = "LPS")
refdata_raw[["treat"]] <- Idents(refdata_raw)

refdata_raw[["id"]] <- "Reference"
refdata_raw[["tech"]] <- "10x"

Seurat_Obj_Diet <- DietSeurat(refdata_raw)
ref.sce <- as.SingleCellExperiment(Seurat_Obj_Diet)

## Cell level QC
per.cell <- perCellQCMetrics(ref.sce, subsets = list(Mito = grep("^mt-", rownames(ref.sce)),
                                                       Ribo = grep("\\bRp[sl]\\d+[^k]*\\b", rownames(ref.sce))))
qc.stats <- quickPerCellQC(per.cell, percent_subsets=c("subsets_Mito_percent","subsets_Ribo_percent"))
qc.stats$high_subsets_Mito_percent <- per.cell$subsets_Mito_percent > 3
qc.stats$discard <- qc.stats$low_lib_size | qc.stats$low_n_features | qc.stats$high_subsets_Mito_percent | qc.stats$high_subsets_Ribo_percent
qc.stats$discard_size_complexity <- qc.stats$low_lib_size | qc.stats$low_n_features
colData(ref.sce) <- cbind(colData(ref.sce), cbind(per.cell, qc.stats))
colSums(as.matrix(qc.stats))
```
```{r}
ref.filtered <- ref.sce[,!ref.sce$discard]

## Feature level QC
keep_feature <- nexprs(ref.filtered, byrow=TRUE) > 0
## Remove mitochondrial and ribosomal genes which does not contribute to analysis (i.e. cell clustering)
keep_feature[grep("\\bRp[sl]\\d+[^k]*\\b", rownames(ref.sce))] <- FALSE
keep_feature[grep("mt-", rownames(ref.sce))] <- FALSE
ref.filtered <- ref.filtered[keep_feature,]
```

```{r}
refdata_filtered <- as.Seurat(ref.filtered)
refdata_filtered[["nCount_originalexp"]] <- NULL; refdata_filtered[["nFeature_originalexp"]] <- NULL
refdata_filtered <- RenameAssays(refdata_filtered, originalexp = "RNA")
refdata_filtered$ident <- NULL

refdata_filtered <- refdata_filtered %>% NormalizeData() %>% FindVariableFeatures() %>% ScaleData() %>% RunPCA() %>% FindNeighbors(dims = 1:50) %>% FindClusters() %>% RunUMAP(dims = 1:50)

load(file = '../data/AAV_MG_scRNAseq/mg_keep_refdat.rda')
```

```{r}
# refdata_raw <- refdata_raw %>% NormalizeData() %>% FindVariableFeatures() %>% ScaleData() %>% RunPCA() %>% FindNeighbors(dims = 1:50) %>% FindClusters() %>% RunUMAP(dims = 1:50)
# 
# d1 <- DimPlot(refdata_raw, cells.highlight = colnames(ref.sce)[ref.sce@colData$discard]) + 
#   NoLegend() + ggtitle("Discard")
# d2 <- DimPlot(refdata_raw, cells.highlight = colnames(ref.sce)[ref.sce@colData$low_lib_size]) + 
#   NoLegend() + ggtitle("low lib size")
# d3 <- DimPlot(refdata_raw, cells.highlight = colnames(ref.sce)[ref.sce@colData$low_n_features]) + 
#   NoLegend() + ggtitle("low n features")
# d4 <- DimPlot(refdata_raw, cells.highlight = colnames(ref.sce)[ref.sce@colData$high_subsets_Mito_percent]) + 
#   NoLegend() + ggtitle("high Mito %")
# d5 <- DimPlot(refdata_raw, cells.highlight = colnames(ref.sce)[ref.sce@colData$high_subsets_Ribo_percent]) + 
#   NoLegend() + ggtitle("high Ribo %")
# 
# (d1 | d2 | d3) / (plot_spacer() | d4 | d5)
```

#### Supplementary Figure 2a

```{r}
as.data.frame(colData(ref.sce)[,c("nCount_RNA","nFeature_RNA","discard_size_complexity")]) %>%
  ggplot(mapping = aes(x=nCount_RNA, y=nFeature_RNA)) +
  geom_point(aes(color = discard_size_complexity), size = 1.2) +
  # ggtitle("Reference Data (10x)") +
  scale_color_manual(values = brewer.pal(n = 6,name = "Paired")[c(2,6)], labels = c("Kept","Removed"), name = "") +
  scale_x_continuous(expand = c(0.15,0.1)) +
  xlab("UMI counts") +
  ylab("Gene counts") +
  theme_classic() +
  theme(
    # plot.title = element_text(size = 12, face = "bold"),
    axis.text = element_text(size = 16),
    axis.title = element_text(size = 21),
    legend.title = element_blank(),
    legend.box.background = element_blank(),
    legend.position = c(.95, .05),
    legend.justification = c("right", "bottom"),
    legend.text = element_text(size = 16),
    legend.margin = margin(1, 5, 5, 5)
    )
ggsave(filename = "refdataQC_numi_ngene.eps", 
       device = "eps", 
       path = "../outs/Figsupp2abcd/", 
       width = 4, height = 3.6, 
       family = "Arial")
```

#### Supplementary Figure 2b

```{r}
as.data.frame(colData(ref.sce)[,c("treat","subsets_Mito_percent","high_subsets_Mito_percent")]) %>%
  ggplot(mapping = aes(x=treat, y=subsets_Mito_percent)) +
  geom_violin() +
  geom_jitter(mapping = aes(color = high_subsets_Mito_percent), size = 0.3, height = 0, width = 0.1) +
  scale_color_manual(values = brewer.pal(n = 6,name = "Paired")[c(2,6)]) +
  xlab(NULL) +
  ylab("Mitochondrial RNA %") +
  theme_linedraw() +
  theme(
    legend.position = "none",
    axis.text.y = element_text(size = 14),
    axis.text.x = element_text(size = 16),
    axis.title = element_text(size = 20),
    )
ggsave(filename = "refdataQC_vln_pctmito.eps", 
       device = "eps", 
       path = "../outs/Figsupp2abcd/", 
       width = 3.2, height = 3.6, 
       family = "Arial")
```

#### Supplementary Figure 2c

```{r}
as.data.frame(colData(ref.sce)[,c("treat","subsets_Ribo_percent","high_subsets_Ribo_percent")]) %>%
  ggplot(mapping = aes(x=treat, y=subsets_Ribo_percent)) +
  geom_violin() +
  geom_jitter(mapping = aes(color = high_subsets_Ribo_percent), size = 0.3, height = 0, width = 0.1) +
  scale_color_manual(values = brewer.pal(n = 6,name = "Paired")[c(2,6)]) +
  xlab(NULL) +
  ylab("Ribosomal RNA %") +
  theme_linedraw() +
  theme(
    legend.position = "none",
    axis.text.y = element_text(size = 14),
    axis.text.x = element_text(size = 16),
    axis.title = element_text(size = 20),
    )
ggsave(filename = "refdataQC_vln_pctribo.eps", 
       device = "eps", 
       path = "../outs/Figsupp2abcd/", 
       width = 3.2, height = 3.6, 
       family = "Arial")
```

#### Supplementary Figure 2d

```{r}
refdata$treat <- factor(refdata$treat, levels = c("Saline","Poly(i:c)","LPS"))
umap_reff1 <- FeaturePlot(refdata_filtered, features = "Hexb") +
  ggtitle("Hexb") +
  theme(
    plot.title = element_text(hjust = 0, size = 24),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    axis.title = element_text(size = 18),
    axis.line = element_line(arrow = grid::arrow(length = unit(0.3, "cm"), 
                                                       ends = "last", type = "closed")),
    legend.title = element_blank(),
    legend.text = element_text(size = 16),
    legend.position = c(.05, .20)
  )
umap_reff2 <- DimPlot(refdata_filtered, cells.highlight = mg_keep_refdat) + 
  NoLegend() +
  ggtitle("Retained Cells") +
  theme(
    plot.title = element_text(hjust = 0, size = 24),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    axis.title = element_text(size = 18),
    axis.line = element_line(arrow = grid::arrow(length = unit(0.3, "cm"), 
                                                 ends = "last", type = "closed"))
  )
umap_reff3 <- DimPlot(refdata, group.by = "treat") +
  scale_color_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,3,5)], 
                     labels = c("Saline", "Poly(i:c)", "LPS")) +
  ggtitle("Final Reference") +
  guides(color = guide_legend(override.aes = list(size=3), byrow = TRUE)) +
  theme(
    plot.title = element_text(hjust = 0, size = 24),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    axis.title = element_text(size = 18),
    axis.line = element_line(arrow = grid::arrow(length = unit(0.3, "cm"), 
                                                 ends = "last", type = "closed")),
    legend.title = element_blank(),
    legend.text = element_text(size = 20),
    legend.position = c(.05, .18),
    legend.spacing.y = unit(0.25, 'cm')
  )

umap_reff1 | umap_reff2 | umap_reff3
ggsave(filename = "umap_refdataQC_retained.eps", 
       device = "eps", 
       path = "../outs/Figsupp2abcd/", 
       width = 15, height = 5, 
       family = "Arial")
```

### Query Data (Smart-seq2)

```{r}
data_long <- read.table(file = '../data/AAV_MG_scRNAseq/AAV_Microglia_counts_all.tsv.gz', sep = "\t", header = TRUE)
data <- pivot_wider(data_long, names_from = "cell", values_from = "count", values_fill = 0)  # Expression matrix

# Optional: Rename rows
anno <- read.table(file = '../misc/MM.GRCm38.102.annotation.tab', sep = "\t", col.names = c("ensembl_id", "symbol"))

data <- data %>% column_to_rownames(var = "gene") %>% as.data.frame()
data <- scWidgets::renameRows(data, anno)

ms <- data.frame(t(data["mScarlet",])); colnames(ms) <- "mScarlet"
data <- data[-which(rownames(data) == "mScarlet"),]

querydata_raw <- CreateSeuratObject(data, project = "MG_AAV_QUERY", names.field = 2, names.delim = "_", meta.data = ms)
querydata_raw$batch <- querydata_raw$orig.ident
querydata_raw$id <- "Query"
querydata_raw$tech <- "Smart-seq2"

Idents(querydata_raw) <- "orig.ident"
querydata_raw <- RenameIdents(querydata_raw, 
                          `yt-8-5` = "AAV-MG1.1", `yt-8-5-2` = "AAV-MG1.1", `yt-8-5-3` = "AAV-MG1.1",
                          `yt-8-5-4` = "AAV-MG1.1", `yt-8-5-6` = "AAV-MG1.1", `yt-8-5-5` = "AAV-MG1.1",
                          `yt-8-6-1` = "AAV-MG1.2", `yt-8-6-2` = "AAV-MG1.2", `yt-8-6-3` = "AAV-MG1.2", 
                          `yt-8-6-4` = "AAV-MG1.2", `yt-8-6-5` = "AAV-MG1.2", `yt-8-6-6` = "AAV-MG1.2",
                          `yt-mgl-L1` = "LPS", `yt-mgl-L2` = "LPS", `yt-mgl-L3` = "LPS")
querydata_raw[["treatment"]] <- Idents(querydata_raw)
querydata_raw[["treatment"]] <- factor(querydata_raw$treatment, 
                                   levels = c("AAV-MG1.2", "AAV-MG1.1","LPS"))
```
```{r}
querydata_raw <- querydata_raw %>% NormalizeData() %>% FindVariableFeatures() %>% ScaleData() %>% RunPCA() %>%
  FindNeighbors(dims = 1:25) %>% FindClusters() %>% RunUMAP(dims = 1:25)
```
```{r}
# querydata_raw <- DietSeurat(querydata_raw)
query.sce <- as.SingleCellExperiment(querydata_raw)
# query.sce <- logNormCounts(query.sce)

## Cell level QC
per.cell <- perCellQCMetrics(query.sce,
                             subsets = list(Mito = grep("^mt-", rownames(query.sce)),
                                            Ribo = grep("\\bRp[sl]\\d+[^k]*\\b", rownames(query.sce))))

qc.stats <- quickPerCellQC(per.cell, percent_subsets = c("subsets_Mito_percent","subsets_Ribo_percent"))
qc.stats$discard_size_complexity <- qc.stats$low_lib_size | qc.stats$low_n_features

qc.stats$low_Hexb <- assay(query.sce, i = "logcounts")["Hexb",] < 2
qc.stats$discard <- qc.stats$low_lib_size | qc.stats$low_n_features | qc.stats$high_subsets_Mito_percent | qc.stats$high_subsets_Ribo_percent | qc.stats$low_Hexb
qc.stats$discard_size_complexity <- qc.stats$low_lib_size | qc.stats$low_n_features

colData(query.sce) <- cbind(colData(query.sce), cbind(per.cell, qc.stats))
colSums(as.matrix(qc.stats))
```
```{r}
query.filtered <- query.sce[,!qc.stats$discard]

## Feature level QC
keep_feature <- nexprs(query.filtered, byrow=TRUE) > 0
## Remove mitochondrial and ribosomal genes which does not contribute to analysis (i.e. cell clustering)
keep_feature[grep("\\bRp[sl]\\d+[^k]*\\b", rownames(query.sce))] <- FALSE
keep_feature[grep("mt-", rownames(query.sce))] <- FALSE
query.filtered <- query.filtered[keep_feature,]

# querydata <- as.Seurat(query.filtered)
```

```{r}
# querydata_raw <- querydata_raw %>% NormalizeData() %>% FindVariableFeatures() %>% ScaleData() %>% RunPCA() %>% FindNeighbors(dims = 1:25) %>% FindClusters() %>% RunUMAP(dims = 1:25)
# 
# d1 <- DimPlot(querydata_raw, cells.highlight = colnames(query.sce)[query.sce@colData$discard]) + 
#   NoLegend() + ggtitle("Discard")
# d2 <- DimPlot(querydata_raw, cells.highlight = colnames(query.sce)[query.sce@colData$low_lib_size]) + 
#   NoLegend() + ggtitle("low lib size")
# d3 <- DimPlot(querydata_raw, cells.highlight = colnames(query.sce)[query.sce@colData$low_n_features]) + 
#   NoLegend() + ggtitle("low n features")
# d4 <- DimPlot(querydata_raw, cells.highlight = colnames(query.sce)[query.sce@colData$high_subsets_Mito_percent]) + 
#   NoLegend() + ggtitle("high Mito %")
# d5 <- DimPlot(querydata_raw, cells.highlight = colnames(query.sce)[query.sce@colData$high_subsets_Ribo_percent]) + 
#   NoLegend() + ggtitle("high Ribo %")
# 
# (d1 | d2 | d3) / (plot_spacer() | d4 | d5)
```

#### Supplementary Figure 2e

```{r}
as.data.frame(colData(query.sce)[,c("nCount_RNA","nFeature_RNA","discard_size_complexity")]) %>%
  ggplot(mapping = aes(x=nCount_RNA, y=nFeature_RNA)) +
  geom_point(aes(color = discard_size_complexity), size = 1.2) +
  # ggtitle("Reference Data (10x)") +
  scale_color_manual(values = brewer.pal(n = 6,name = "Paired")[c(2,6)], labels = c("Kept","Removed"), name = "") +
  scale_x_continuous(expand = c(0.15,0.1)) +
  xlab("UMI counts") +
  ylab("Gene counts") +
  theme_classic() +
  theme(
    # plot.title = element_text(size = 12, face = "bold"),
    axis.text = element_text(size = 16),
    axis.title = element_text(size = 21),
    legend.title = element_blank(),
    legend.box.background = element_blank(),
    legend.position = c(.95, .05),
    legend.justification = c("right", "bottom"),
    legend.text = element_text(size = 16),
    legend.margin = margin(1, 5, 5, 5)
    )
ggsave(filename = "querydataQC_numi_ngene.eps", 
       device = "eps", 
       path = "../outs/Figsupp2efgh/", 
       width = 4, height = 3.6, 
       family = "Arial")
```

#### Supplementary Figure 2f

```{r}
as.data.frame(colData(query.sce)[,c("treatment","subsets_Mito_percent","high_subsets_Mito_percent")]) %>%
  mutate(treatment = factor(treatment, levels = c("AAV-MG1.1","AAV-MG1.2","LPS"))) %>%
  ggplot(mapping = aes(x=treatment, y=subsets_Mito_percent)) +
  geom_violin() +
  geom_jitter(mapping = aes(color = high_subsets_Mito_percent), size = 0.3, height = 0, width = 0.1) +
  scale_x_discrete(labels = c("AAV-\nMG-1.1","AAV-\nMG-1.2","LPS")) +
  scale_color_manual(values = brewer.pal(n = 6,name = "Paired")[c(2,6)]) +
  xlab(NULL) +
  ylab("Mitochondrial RNA %") +
  theme_linedraw() +
  theme(
    legend.position = "none",
    axis.text.y = element_text(size = 14),
    axis.text.x = element_text(size = 16),
    axis.title = element_text(size = 20),
    )
ggsave(filename = "querydataQC_vln_pctmito.eps", 
       device = "eps", 
       path = "../outs/Figsupp2efgh/", 
       width = 3.2, height = 3.6, 
       family = "Arial")
```

#### Supplementary Figure 2g

```{r}
as.data.frame(colData(query.sce)[,c("treatment","subsets_Ribo_percent","high_subsets_Ribo_percent")]) %>%
  mutate(treatment = factor(treatment, levels = c("AAV-MG1.1","AAV-MG1.2","LPS"))) %>%
  ggplot(mapping = aes(x=treatment, y=subsets_Ribo_percent)) +
  geom_violin() +
  geom_jitter(mapping = aes(color = high_subsets_Ribo_percent), size = 0.3, height = 0, width = 0.1) +
  scale_x_discrete(labels = c("AAV-\nMG-1.1","AAV-\nMG-1.2","LPS")) +
  scale_color_manual(values = brewer.pal(n = 6,name = "Paired")[c(2,6)]) +
  xlab(NULL) +
  ylab("Ribosomal RNA %") +
  theme_linedraw() +
  theme(
    legend.position = "none",
    axis.text.y = element_text(size = 14),
    axis.text.x = element_text(size = 16),
    axis.title = element_text(size = 20),
    )
ggsave(filename = "querydataQC_vln_pctribo.eps", 
       device = "eps", 
       path = "../outs/Figsupp2efgh/", 
       width = 3.2, height = 3.6, 
       family = "Arial")
```

#### Supplementary Figure 2h

```{r}
umap_queryf1 <- FeaturePlot(querydata_raw, features = "Hexb") +
  ggtitle("Hexb") +
  theme(
    plot.title = element_text(hjust = 0, size = 24),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    axis.title = element_text(size = 18),
    axis.line = element_line(arrow = grid::arrow(length = unit(0.3, "cm"), 
                                                       ends = "last", type = "closed")),
    legend.title = element_blank(),
    legend.text = element_text(size = 16),
    legend.position = c(.05, .20)
  )
umap_queryf2 <- DimPlot(querydata_raw, cells.highlight = colnames(query.sce)[!query.sce@colData$discard]) + 
  NoLegend() + 
  ggtitle("Retained Cells") +
  theme(
    plot.title = element_text(hjust = 0, size = 24),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    axis.title = element_text(size = 18),
    axis.line = element_line(arrow = grid::arrow(length = unit(0.3, "cm"), 
                                                 ends = "last", type = "closed"))
  )
umap_queryf3 <- DimPlot(querydata, group.by = "treatment") +
  scale_color_manual(values = brewer.pal(n = 12, name = "Paired")[c(12,8,6)],
                     labels = c("AAV-MG1.1", "AAV-MG1.2", "LPS")) +
  ggtitle("Final Query") +
  guides(color = guide_legend(override.aes = list(size=3), byrow = TRUE)) +
  theme(
    plot.title = element_text(hjust = 0, size = 24),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    axis.title = element_text(size = 18),
    axis.line = element_line(arrow = grid::arrow(length = unit(0.3, "cm"), 
                                                 ends = "last", type = "closed")),
    legend.title = element_blank(),
    legend.text = element_text(size = 20),
    legend.position = c(.55, .18),
    legend.spacing.y = unit(0.25, 'cm')
  )

umap_queryf1 | umap_queryf2 | umap_queryf3
ggsave(filename = "umap_querydataQC_retained.eps", 
       device = "eps", 
       path = "../outs/Figsupp2efgh/", 
       width = 15, height = 5, 
       family = "Arial")

```

#### Supplementary Figure 2i

```{r}
refquery$id <- factor(refquery$id, levels = c("Reference","Query"))
```
```{r}
as.data.frame(refquery[[c("nCount_RNA","nFeature_RNA","id")]]) %>%
  ggplot(mapping = aes(x=id, y=nCount_RNA)) +
  geom_violin() +
  geom_jitter(aes(color = id), size = 0.3, height = 0, width = 0.3) +
  scale_x_discrete(labels = c("Reference\n(10x)","Query\n(Smart-seq2)")) +
  scale_y_continuous(expand = c(0.1,0.1)) +
  xlab(NULL) +
  ylab("UMI counts") +
  theme_linedraw() +
  theme(
    legend.position = "none",
    axis.text.x = element_text(size = 16),
    axis.text.y = element_text(size = 14),
    axis.title.y =  element_text(size = 18)
    )
ggsave(filename = "vln_numi.eps", 
       device = "eps", 
       path = "../outs/Figsupp2ij/", 
       width = 3.2, height = 3.6, 
       family = "Arial")
```

#### Supplementary Figure 2j

```{r}
as.data.frame(refquery[[c("nFeature_RNA","id")]]) %>%
  ggplot(mapping = aes(x=id, y=nFeature_RNA)) +
  geom_violin() +
  geom_jitter(aes(color = id), size = 0.3, height = 0, width = 0.3) +
  scale_x_discrete(labels = c("Reference\n(10x)","Query\n(Smart-seq2)")) +
  scale_y_continuous(expand = c(0.1,0.1)) +
  xlab(NULL) +
  ylab("Genes Detected") +
  theme_linedraw() +
  theme(
    legend.position = "none",
    axis.text.x = element_text(size = 16),
    axis.text.y = element_text(size = 14),
    axis.title.y =  element_text(size = 18)
    )
ggsave(filename = "vln_ngene.eps", 
       device = "eps", 
       path = "../outs/Figsupp2ij/", 
       width = 3.2, height = 3.6, 
       family = "Arial")
```

## DEG analysis and marker expressions (Supplementary Figure 2k-n)

### Reference Dataset (10x)

#### Supplementary Figure 2k

##### DEGs Heatmap

Differential expression analysis. Used subsequently for heatmap plots.
```{r}
Idents(refdata) <- "treat"
mks.refdata.all <- FindAllMarkers(refdata, only.pos = T)
# Idents(querydata) <- "group"
# mks.querydata.all <- FindAllMarkers(querydata, only.pos = T)
Idents(query_transferred) <- "group"
mks.querydata.all <- FindAllMarkers(query_transferred, only.pos = T)
```

Parse DEGs.
```{r}
mkset.1 <- intersect(mks.refdata.all[mks.refdata.all$cluster == "Saline",]$gene, 
                     mks.querydata.all[mks.querydata.all$cluster == "AAV",]$gene)

mkset.2 <- mks.refdata.all[mks.refdata.all$cluster == "Poly(i:c)",]$gene

mkset.3 <- intersect(mks.refdata.all[mks.refdata.all$cluster %in% c("LPS","Poly(i:c)"),]$gene, 
                     mks.querydata.all[mks.querydata.all$cluster == "LPS",]$gene)
```

Preparing for heatmap plot.
```{r}
refdata$treat <- factor(refdata$treat, levels = c("Saline","Poly(i:c)","LPS"))
refdata <- ScaleData(refdata, features = c(mkset.1, mkset.2, mkset.3, VariableFeatures(refdata)))
```

Heatmap of DEGs in Saline, Poly(i:c), LPS groups in reference dataset.
```{r}
## Downsample for demo
DoHeatmap(subset(refdata, downsample = 500),
          features = c(mkset.1, mkset.2, mkset.3),
          group.by = "treat",
          group.colors = brewer.pal(n = 12, name = "Paired")[c(1,3,5)]) +
  scale_fill_distiller(palette = "RdBu") +
  theme(
    axis.text.y = element_blank()
    )
  
```
```{r}
## Extremely slow, causes Rstudio to crash. Only run if necessary.
DoHeatmap(refdata,
          features = c(mkset.1, mkset.2, mkset.3),
          group.by = "treat",
          group.colors = brewer.pal(n = 12, name = "Paired")[c(1,3,5)],
          label = FALSE) +
  scale_fill_distiller(palette = "RdBu") +
  theme(
    axis.text.y = element_blank()
    ) +
  NoLegend()
  
```

##### Violin plots of Representative Genes

```{r}
Idents(refdata) <- "treat"
features_hom = c("P2ry12","P2ry13","Cx3cr1","Csf1r","Lrp1","Ptpre","Lipa","Ets1","Sall1","Sirpa")
features_hom_lab = c( "Csf1r")

vplots = list()
for (feature in features_hom){
  vplots[[feature]] <- FetchData(refdata, vars = c("treat", feature)) %>% 
    ggplot(aes_string(x = "treat", y = feature)) + 
    geom_violin(aes_string(fill = "treat")) + 
    geom_jitter(aes_string(color = "treat"), size = 0.3, alpha = 0.2, width = 0.25) + 
    scale_color_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,3,5)]) +
    scale_fill_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,3,5)]) +
    theme_classic() +
    theme(
      legend.position = "none",
      axis.title.x = element_blank(),
      axis.text.x = element_blank(),
      axis.title.y = element_text(size = 21, family = "Arial"),
      axis.text.y = element_text(size = 16, family = "Arial")
    )
}

for (feature in features_hom_lab){
  vplots[[feature]] <- FetchData(refdata, vars = c("treat", feature)) %>% 
    ggplot(aes_string(x = "treat", y = feature)) + 
    geom_violin(aes_string(fill = "treat")) + 
    geom_jitter(aes_string(color = "treat"), size = 0.3, alpha = 0.2, width = 0.25) + 
    scale_color_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,3,5)]) +
    scale_fill_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,3,5)]) +
    theme_classic() +
    theme(
      legend.position = "none",
      axis.title.x = element_blank(),
      axis.text.x = element_text(size = 20, family = "Arial"),
      axis.title.y = element_text(size = 21, family = "Arial"),
      axis.text.y = element_text(size = 16, family = "Arial")
    )
}

# (vplots[["P2ry12"]] | vplots[["P2ry13"]]) / (vplots[["Cx3cr1"]] | vplots[["Csf1r"]]) / (vplots[["Sall1"]] | vplots[["Sirpa"]])
# ggsave(filename = "ref_hom_vlns.tiff",
#        device = "tiff",
#        path = "../outs/Figsupp2kl/",
#        width = 5, height = 4.5)

## In order to preserve transparency, use cairo_ps to generate eps plot.
## Don't run in code chunk. Run in console.
## Weird jitter dots with transparency enabled (alpha = 0.2), but fonts looks much better.
cairo_ps(filename = 'outs/Figsupp2kl/ref_hom_vlns.eps', width = 4, height = 6.4)
vplots[["P2ry12"]] / vplots[["Cx3cr1"]] / vplots[["Sall1"]] / vplots[["Csf1r"]]
dev.off()
```

```{r}
features_rea = c("Stat3","Gbp7","Ptgds","Irf7","Irf9","Saa3","Tlr2","Mt1","Mt2","Cd74","Lcn2","Tnf","Il1b","Cst7","Cxcl10",
                 "Ifitm3","Spp1")
features_rea_lab = c("Ccl5")

vplots = list()
for (feature in features_rea){
  vplots[[feature]] <- FetchData(refdata, vars = c("treat", feature)) %>% 
    ggplot(aes_string(x = "treat", y = feature)) + 
    geom_violin(aes_string(fill = "treat")) + 
    geom_jitter(aes_string(color = "treat"), size = 0.3, alpha = 0.2, width = 0.25) + 
    scale_color_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,3,5)]) +
    scale_fill_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,3,5)]) +
    theme_classic() +
    theme(
      legend.position = "none",
      axis.title.x = element_blank(),
      axis.text.x = element_blank(),
      axis.title.y = element_text(size = 21, family = "Arial"),
      axis.text.y = element_text(size = 16, family = "Arial")
    )
}

for (feature in features_rea_lab){
  vplots[[feature]] <- FetchData(refdata, vars = c("treat", feature)) %>% 
    ggplot(aes_string(x = "treat", y = feature)) + 
    geom_violin(aes_string(fill = "treat")) + 
    geom_jitter(aes_string(color = "treat"), size = 0.3, alpha = 0.2, width = 0.25) + 
    scale_color_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,3,5)]) +
    scale_fill_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,3,5)]) +
    theme_classic() +
    theme(
      legend.position = "none",
      axis.title.x = element_blank(),
      axis.text.x = element_text(size = 20, family = "Arial"),
      axis.title.y = element_text(size = 21, family = "Arial"),
      axis.text.y = element_text(size = 16, family = "Arial")
    )
}

# (vplots[["Tnf"]] | vplots[["Il1b"]]) / (vplots[["Cd74"]] | vplots[["Cst7"]]) / (vplots[["Spp1"]] | vplots[["Irf9"]])/ (vplots[["Ccl2"]] | vplots[["Ccl5"]])
# ggsave(filename = "ref_rea_vlns.tiff",
#        device = "tiff",
#        path = "../outs/Figsupp2kl/",
#        width = 5, height = 6)

cairo_ps(filename = 'outs/Figsupp2kl/ref_rea_vlns.eps', width = 4, height = 6.4)
vplots[["Il1b"]] / vplots[["Spp1"]] / vplots[["Ifitm3"]] / vplots[["Ccl5"]]
dev.off()
```

### Query Dataset (Smart-seq2)

#### Supplementary Figure 2l

##### DEGs Heatmap

Preparing for heatmap plot.
```{r}
query_transferred$ident <- factor(query_transferred$ident, levels = c("Non-transduced","Transduced","LPS(Query)"))
query_transferred <- ScaleData(query_transferred, features = c(mkset.1, mkset.2, mkset.3, VariableFeatures(query_transferred)))
```

Heatmap of DEGs in AAV, LPS groups in query dataset.
```{r}
## Demo
DoHeatmap(query_transferred,
          features = c(mkset.1,mkset.3),
          group.by = "ident",
          group.colors = brewer.pal(n = 12, name = "Paired")[c(1,8,6)]) +
  theme(
    axis.text.y = element_blank()
    ) +
  scale_fill_distiller(palette = "PiYG")
```
```{r}
DoHeatmap(query_transferred,
          features = c(mkset.1,mkset.3),
          group.by = "ident",
          group.colors = brewer.pal(n = 12, name = "Paired")[c(1,8,6)],
          label = FALSE) +
  theme(
    axis.text.y = element_blank()
    ) +
  scale_fill_distiller(palette = "PiYG") +
  NoLegend()
```

##### Violin plots of Representative Genes

```{r}
query_transferred$ident <- factor(query_transferred$ident, levels = c("Non-transduced","Transduced","LPS(Query)"))
Idents(query_transferred) <- "ident"
```

```{r}
features_hom = c("P2ry12","P2ry13","Cx3cr1","Csf1r","Lrp1","Ptpre","Lipa","Ets1","Sall1","Sirpa")
features_hom_lab = c( "Csf1r")

vplots = list()
for (feature in features_hom){
  vplots[[feature]] <- FetchData(query_transferred, vars = c("treatment", feature)) %>% 
    ggplot(aes_string(x = "treatment", y = feature)) + 
    geom_jitter(aes_string(color = "treatment"), size = 0.6, alpha = 1, width = 0.25) + 
    geom_violin(aes_string(fill = "treatment"), alpha = 0.6) + 
    scale_color_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,7,5)]) +
    scale_fill_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,7,5)]) +
    theme_classic() +
    theme(
      legend.position = "none",
      axis.title.x = element_blank(),
      axis.text.x = element_blank(),
      axis.title.y = element_text(size = 21, family = "Arial"),
      axis.text.y = element_text(size = 16, family = "Arial")
    )
}

for (feature in features_hom_lab){
  vplots[[feature]] <- FetchData(query_transferred, vars = c("treatment", feature)) %>% 
    ggplot(aes_string(x = "treatment", y = feature)) + 
    geom_jitter(aes_string(color = "treatment"), size = 0.6, alpha = 1, width = 0.25) + 
    geom_violin(aes_string(fill = "treatment"), alpha = 0.6) + 
    scale_color_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,7,5)]) +
    scale_fill_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,7,5)]) +
    scale_x_discrete(labels = c("Untransd.","Transd.","LPS")) +
    theme_classic() +
    theme(
      legend.position = "none",
      axis.title.x = element_blank(),
      axis.text.x = element_text(size = 20, family = "Arial"),
      axis.title.y = element_text(size = 21, family = "Arial"),
      axis.text.y = element_text(size = 16, family = "Arial")
    )
}

# (vplots[["P2ry12"]] | vplots[["P2ry13"]]) / (vplots[["Cx3cr1"]] | vplots[["Csf1r"]]) / (vplots[["Sall1"]] | vplots[["Sirpa"]])
# ggsave(filename = "query_hom_vlns.tiff",
#        device = "tiff",
#        path = "../outs/Figsupp2kl/",
#        width = 5, height = 4.5)

cairo_ps(filename = 'outs/Figsupp2kl/query_hom_vlns.eps', width = 4, height = 6.4)
vplots[["P2ry12"]] / vplots[["Cx3cr1"]] / vplots[["Sall1"]] / vplots[["Csf1r"]]
dev.off()
```

```{r}
features_rea = c("Stat3","Gbp7","Ptgds","Irf7","Irf9","Saa3","Tlr2","Mt1","Mt2","Cd74","Lcn2","Tnf","Il1b","Cst7","Cxcl10",
                 "Ifitm3","Spp1")
features_rea_lab = c("Ccl5")

vplots = list()
for (feature in features_rea){
  vplots[[feature]] <- FetchData(query_transferred, vars = c("treatment", feature)) %>% 
    ggplot(aes_string(x = "treatment", y = feature)) + 
    geom_jitter(aes_string(color = "treatment"), size = 0.6, alpha = 1, width = 0.25) + 
    geom_violin(aes_string(fill = "treatment"), alpha = 0.6) + 
    scale_color_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,7,5)]) +
    scale_fill_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,7,5)]) +
    theme_classic() +
    theme(
      legend.position = "none",
      axis.title.x = element_blank(),
      axis.text.x = element_blank(),
      axis.title.y = element_text(size = 21, family = "Arial"),
      axis.text.y = element_text(size = 16)
    )
}

for (feature in features_rea_lab){
  vplots[[feature]] <- FetchData(query_transferred, vars = c("treatment", feature)) %>% 
    ggplot(aes_string(x = "treatment", y = feature)) + 
    geom_jitter(aes_string(color = "treatment"), size = 0.6, alpha = 1, width = 0.25) + 
    geom_violin(aes_string(fill = "treatment"), alpha = 0.6) + 
    scale_color_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,7,5)]) +
    scale_fill_manual(values = brewer.pal(n = 12, name = "Paired")[c(1,7,5)]) +
    scale_x_discrete(labels = c("Untransd.","Transd.","LPS")) +
    theme_classic() +
    theme(
      legend.position = "none",
      axis.title.x = element_blank(),
      axis.text.x = element_text(size = 20, family = "Arial"),
      axis.title.y = element_text(size = 21, family = "Arial"),
      axis.text.y = element_text(size = 16, family = "Arial")
    )
}

# (vplots[["Tnf"]] | vplots[["Il1b"]]) / (vplots[["Cd74"]] | vplots[["Cst7"]]) / (vplots[["Spp1"]] | vplots[["Irf9"]])/ (vplots[["Ccl2"]] | vplots[["Ccl5"]])
# ggsave(filename = "query_rea_vlns.tiff",
#        device = "tiff",
#        path = "../outs/Figsupp2kl/",
#        width = 5, height = 6)

cairo_ps(filename = 'outs/Figsupp2kl/query_rea_vlns.eps', width = 4, height = 6.4)
vplots[["Il1b"]] / vplots[["Spp1"]] / vplots[["Ifitm3"]] / vplots[["Ccl5"]]
dev.off()
```

